examples/graceful-shutdown/src/main.rs

```diff
@@ -4,57 +4,165 @@
+            TimeoutLayer::new(Duration::from_secs(10)),
+        ));
+
+    // Create a `TcpListener` using tokio.
+    let listener = TcpListener::bind("0.0.0.0:3000").await.unwrap();
+
+    // Create a watch channel to track tasks that are handling connections and wait for them to
+    // complete.
+    let (close_tx, close_rx) = watch::channel(());
+
+    // Continuously accept new connections.
+    loop {
+        let (socket, remote_addr) = tokio::select! {
+            // Either accept a new connection...
+            result = listener.accept() => {
+                result.unwrap()
+            }
+            // ...or wait to receive a shutdown signal and stop the accept loop.
+            _ = shutdown_signal() => {
+                debug!("signal received, not accepting new connections");
+                break;
+            }
+        };
+
+        debug!("connection {remote_addr} accepted");
+
+        // We don't need to call `poll_ready` because `Router` is always ready.
+        let tower_service = app.clone();
+
+        // Clone the watch receiver and move it into the task.
+        let close_rx = close_rx.clone();
+
+        // Spawn a task to handle the connection. That way we can multiple connections
+        // concurrently.
+        tokio::spawn(async move {
+            // Hyper has its own `AsyncRead` and `AsyncWrite` traits and doesn't use tokio.
+            // `TokioIo` converts between them.
+            let socket = TokioIo::new(socket);
 
-// TODO
-fn main() {
-    eprint!("this example has not yet been updated to hyper 1.0");
+            // Hyper also has its own `Service` trait and doesn't use tower. We can use
+            // `hyper::service::service_fn` to create a hyper `Service` that calls our app through
+            // `tower::Service::call`.
+            let hyper_service = hyper::service::service_fn(move |request: Request<Incoming>| {
+                // We have to clone `tower_service` because hyper's `Service` uses `&self` whereas
+                // tower's `Service` requires `&mut self`.
+                //
+                // We don't need to call `poll_ready` since `Router` is always ready.
+                tower_service.clone().call(request)
+            });
+
+            // `hyper_util::server::conn::auto::Builder` supports both http1 and http2 but doesn't
+            // support graceful so we have to use hyper directly and unfortunately pick between
+            // http1 and http2.
+            let conn = hyper::server::conn::http1::Builder::new()
+                .serve_connection(socket, hyper_service)
+                // `with_upgrades` is required for websockets.
+                .with_upgrades();
+
+            // `graceful_shutdown` requires a pinned connection.
+            let mut conn = std::pin::pin!(conn);
+
+            loop {
+                tokio::select! {
+                    // Poll the connection to process requests.
+                    result = conn.as_mut() => {
```

Comment:
The `conn` future only resolves when the clieng has closed the connection, there has been an unrecoverable TCP error, or the connection is idle and graceful shutdown has been requested, I guess? Might be worth extending the comment above.